tokio-util/src/task/spawn_pinned.rs

```diff
@@ -0,0 +1,247 @@
+            let spawn_task = Box::new(move || {
+                // Once we're in the LocalSet context we can call spawn_local,
+                // but first try to create the task. We use catch_unwind here so
+                // we don't kill the worker if create_task panics.
+                let task = match std::panic::catch_unwind(AssertUnwindSafe(create_task)) {
+                    Ok(task) => task,
+                    Err(e) => {
+                        // Creating the task triggered a panic. Send the error
+                        // back so it gets propagated.
+                        let _ = sender.send(Err(e));
+                        return;
+                    }
+                };
+
+                // Task was created successfully, now spawn it.
+                let join_handle = spawn_local(task);
+
+                // Send the join handle back to the spawner.
+                let _ = sender.send(Ok(join_handle));
+            });
+
+            // Send the callback to the LocalSet task
+            if let Err(e) = worker_handle_clone.spawner.send(spawn_task) {
+                // Failed to spawn the task on the worker, so we need to remove
+                // the task from the count.
+                worker_handle_clone.decrement_task_count();
+
+                // Propagate the error as a panic in the join handle.
+                panic!("Failed to send job to worker: {}", e);
+            }
+
+            // Wait for the task's join handle
+            let join_handle = match receiver.await {
+                Ok(Ok(handle)) => handle,
+                Ok(Err(e)) => {
+                    // create_task panicked. Decrement the task count and
+                    // propagate the panic here.
+                    worker_handle_clone.decrement_task_count();
+                    std::panic::resume_unwind(e);
+                }
+                Err(e) => {
+                    // We sent the task successfully, but failed to get its
+                    // join handle... We assume something happened to the worker
+                    // and the task was not spawned, so we need to remove the
+                    // task from the count.
+                    worker_handle_clone.decrement_task_count();
+
+                    // Propagate the error as a panic in the join handle.
+                    panic!("Worker failed to send join handle: {}", e);
+                }
+            };
+
+            // Wait for the task to complete
+            let join_result = join_handle.await;
+
+            // Update the task count once the future has finished
+            worker_handle_clone.decrement_task_count();
+
+            match join_result {
+                Ok(output) => output,
+                Err(e) => {
+                    // Task panicked or was canceled. Forward this error as a
+                    // panic in the join handle.
+                    panic!("spawn_pinned task panicked or aborted: {}", e);
+                }
```

Comment:
The `JoinHandle` has an `into_panic` method you can use together with `resume_unwind`.